diff -Naur openssl-openssl-3.5.0/crypto/cms/cms_pwri.c openssl-changed/crypto/cms/cms_pwri.c
--- openssl-openssl-3.5.0/crypto/cms/cms_pwri.c	2025-04-08 21:09:06.000000000 +0800
+++ openssl-changed/crypto/cms/cms_pwri.c	2025-12-23 19:59:27.652036759 +0800
@@ -242,7 +242,7 @@
         /* Check byte failure */
         goto err;
     }
-    if (inlen < (size_t)(tmp[0] - 4)) {
+    if (inlen < 4 + (size_t)tmp[0]) {
         /* Invalid length value */
         goto err;
     }
diff -Naur openssl-openssl-3.5.0/crypto/ec/ecp_sm2p256.c openssl-changed/crypto/ec/ecp_sm2p256.c
--- openssl-openssl-3.5.0/crypto/ec/ecp_sm2p256.c	2025-04-08 21:09:06.000000000 +0800
+++ openssl-changed/crypto/ec/ecp_sm2p256.c	2025-12-23 20:01:15.230505034 +0800
@@ -747,7 +747,7 @@
         ossl_ec_GFp_simple_point_copy,
         ossl_ec_GFp_simple_point_set_to_infinity,
         ossl_ec_GFp_simple_point_set_affine_coordinates,
-        ecp_sm2p256_get_affine,
+        ossl_ec_GFp_simple_point_get_affine_coordinates,
         0, 0, 0,
         ossl_ec_GFp_simple_add,
         ossl_ec_GFp_simple_dbl,
@@ -763,7 +763,7 @@
         ecp_sm2p256_field_mul,
         ecp_sm2p256_field_sqr,
         0 /* field_div */,
-        0 /* field_inv */,
+        ossl_ec_GFp_simple_field_inv,
         0 /* field_encode */,
         0 /* field_decode */,
         0 /* field_set_to_one */,
@@ -779,7 +779,7 @@
         ossl_ecdsa_simple_sign_setup,
         ossl_ecdsa_simple_sign_sig,
         ossl_ecdsa_simple_verify_sig,
-        ecp_sm2p256_inv_mod_ord,
+        0, /* use constant‑time fallback for inverse mod order */
         0, /* blind_coordinates */
         0, /* ladder_pre */
         0, /* ladder_step */
diff -Naur openssl-openssl-3.5.0/crypto/http/http_lib.c openssl-changed/crypto/http/http_lib.c
--- openssl-openssl-3.5.0/crypto/http/http_lib.c	2025-04-08 21:09:06.000000000 +0800
+++ openssl-changed/crypto/http/http_lib.c	2025-12-23 20:03:53.412604285 +0800
@@ -269,6 +269,7 @@
         /* strip leading '[' and trailing ']' from escaped IPv6 address */
         sl -= 2;
         strncpy(host, server + 1, sl);
+        host[sl] = '\0';
         server = host;
     }
 
